logging/common/src/main/java/io/helidon/logging/common/BaseMdcProvider.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2025 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.logging.common;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import io.helidon.logging.common.spi.MdcProvider;
+
+/**
+ * Implements {@link io.helidon.logging.common.spi.MdcProvider} using an internal thread local data structure.
+ */
+public class BaseMdcProvider implements MdcProvider {
```

Comment:
This class should be abstract.

Reply:
I think I see a way of reorganizing the code a bit that avoids duplicating code while addressing your comments. I'll do that.

That said, you didn't describe what problems you see this introducing.

I'm not sure how there could be one `MdcProvider` system-wide; the `HelidonMdc` code specifically allows for and invokes however many providers are found by service loading.